### PR TITLE
Improve of docker-compose multisrv

### PR DESCRIPTION
**Pull request by mliradelc**
This is an improvement to the docker compose script made by @mtneug add some explicit declaration and improvements for the database. This PR solves #109 and #110. 

- Database Changes
  - Change of SQL to Opencast 7.x script.
~~- SQL entry point explicit declaration~~
  - Increase of database wait timeout from 600 to 28800

~~- Network Changes~~
~~- Aliases for docker network now with an explicit declaration~~
 ~~- Port 3306 no declared outside to avoid host conflicts.~~

~~Docker Initialization~~
~~- Opencast nodes will wait for ActiveMQ and MariaDB containers to be ready before they start to be deployed.~~

### CODE REVIEW

**Comment by mliradelc** [comment]
Ok, this is unexpected, Travis is having trouble building Opencast and I don't have an idea why the files I modify are not supposed to be built.

**Comment by mtneug** [comment]
This is unrelated to your change. This error started showing up a couple of days ago. My guess is that either the base image or something with the dependencies was changed. I will look into it today.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 1]
```diff
@@ -13,9 +13,11 @@
 # limitations under the License.
 
 version: "3"
+
 volumes:
   data: {}
   db: {}
+  assets:
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: This creates a new volume named `assets`, but it is never used. The `./assets/` volume is local on the filesystem and not directly managed by Docker. This should be removed.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 2]
```diff
@@ -62,8 +73,15 @@ services:
       - ACTIVEMQ_BROKER_PASSWORD=password
     ports:
       - "8080:8080"
+    depends_on:
+      - mariadb
+      - activemq
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: While this will ensure that the Opencast containers are only started after MariaDB and ActiveMQ, it will not ensure that those two are ready for Opencast to connect to. The Opencast image has a mechanism to wait for the DB connection and Opencast itself waits for ActiveMQ. This is therefore unnecessary and has not the desired effect anyway.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 3]
```diff
@@ -24,11 +26,16 @@ services:
...
       - MYSQL_USER=opencast
       - MYSQL_PASSWORD=opencast
-    ports:
-      - "3306:3306"
     volumes:
       - ./assets/opencast-ddl.sql:/docker-entrypoint-initdb.d/opencast-ddl.sql:ro
       - db:/var/lib/mysql
+    command: [
+            '--wait_timeout=28800',
+        ]
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Could you use the more compact version: `command: "--wait_timeout=28800"`

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 4]
```diff
@@ -62,8 +73,15 @@ services:
       - ACTIVEMQ_BROKER_PASSWORD=password
     ports:
       - "8080:8080"
+    depends_on:
+      - mariadb
+      - activemq
     volumes:
-      - data:/data
+      - ./data:/data
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Why the change to a local mount?

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 5]
```diff
@@ -24,11 +26,16 @@ services:
...
     volumes:
       - ./assets/opencast-ddl.sql:/docker-entrypoint-initdb.d/opencast-ddl.sql:ro
       - db:/var/lib/mysql
+    command: [
+            '--wait_timeout=28800',
+        ]
+    networks:
+      default:
+        aliases:
+          - mariadb
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: I don't like this change. This hostname should already be set by default as the service is named `mariadb` (see e.g. <https://docs.docker.com/compose/networking/>). If this is not working for you, this is either a bug (unlikely I have the exact same version running on a Mac) or something is up with your setup.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 6]
```diff
@@ -24,11 +26,16 @@ services:
       - MYSQL_DATABASE=opencast
       - MYSQL_USER=opencast
       - MYSQL_PASSWORD=opencast
-    ports:
-      - "3306:3306"
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: No strong feelings here. I published the port just for connecting to the DB locally. It is probably better to remove it as this should not be published in production.

File: docker-compose/assets/opencast-ddl.sql [hunk 7]
```diff
@@ -1,5 +1,3 @@
--- Created with Opencast version next
-
```
Comment (mtneug) [inline docker-compose/assets/opencast-ddl.sql]: Could you add this comment? There is a mechanism to generate the DDL script from a given Opencast image (see README.md) and this comment is always included to indicate the version.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 1 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Yes, this was a solution I found when the database entry point was not executed.

Then after making more tests and reading the log files, it was de DDL file was not the correct.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 2 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Even so, the containers have built their processes to wait for the database and the message broker, this helps to declutter the Logfile and ensures good practice.

File: docker-compose/assets/opencast-ddl.sql (hunk 7 above)
Comment (mliradelc) [reply docker-compose/assets/opencast-ddl.sql]: If its creating the DDL script given the opencast image is not working properly, I have to manually set the DDL from opencast version 7 because the original script have some tables missing

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 6 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: If you have the same port outside creates a possible problem if you have a MySQL running from before, the thing I can do is put another exit port or left without exit. ;)

**Comment by mtneug** [comment]
Could you also rebase to latest commit? CI should then succeed again.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 6 above)
Comment (mtneug) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Again this is fine by me. Just wanting to explain the reason for publishing the port ;)

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 3 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Solved

File: docker-compose/assets/opencast-ddl.sql (hunk 7 above)
Comment (mtneug) [reply docker-compose/assets/opencast-ddl.sql]: The image does not automatically use the DDL script (well Opencast can). What I mean is that given an Opencast image Docker you can run this command
```sh
$ docker run -it --rm \
  -e ORG_OPENCASTPROJECT_DB_VENDOR=MySQL \
  quay.io/opencast/allinone:next \
  app:print:ddl
```
to print out the DDL script necessary for this Opencast version to work. The output will include the above header to indicate which version of Opencast this script is tight to. Therefore I was asking to add it again.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 4 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: It gives you the advance to look into the events files, but for this case is not needed.

I rolled back to the original statement.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 5 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: It was problem with the DDL file, rolling back this change

**Comment by mliradelc** [comment]
> Could you also rebase to latest commit? CI should then succeed again.

Thanks, I added your latest changes.

The biggest issue was the problem with the timeout and the DDL file at the end. You said that the DDL is generated, What is the advantage to use that and put the original opencast DDL file?

And the port, I suggest to leave it closed (Or with an alternative host port).

File: docker-compose/assets/opencast-ddl.sql (hunk 7 above)
Comment (mliradelc) [reply docker-compose/assets/opencast-ddl.sql]: Ok, now I understand better. It was for me very strange that comment because says "Created" like generated from a Script.

So is not necessary when you use Compose but you need that comment when you run as a standalone image. :) Have I understood correctly?

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 2 above)
Comment (mtneug) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: I disagree on the "good practice" part for the following reasons:

1. This does not solve the dependency problem as it does not guarantee that the downstream services are ready (see <https://docs.docker.com/v17.09/compose/startup-order/>).
1. Docker Swarm [completely ignores this setting](https://docs.docker.com/v17.09/compose/compose-file/#depends_on). In these environments the idea is that your application is resilient enough to handle temporary connection losses and arbitrary startup orders.
1. Each Opencast image therefore first checks if the DB is available with the help of a [little Java program](https://github.com/opencast/opencast-docker/blob/a8fe650006b5926edbb718954a8c6e9240955136/Dockerfiles/allinone/assets/scripts/TryToConnectToDb.java) and only then continues to start Opencast. For ActiveMQ this is handled directly by Opencast. This should be the fastest way to boot up the cluster.

I'm therefore not a fan of the `depends_on` option and I'm not recommending using it.

I can see that it at least documents the dependencies and maybe it leads to decluttered logs during startup (you can also get the logs for each service with `docker-compose logs -f <myservice>`). But do you think this is enough of a justification? As this is an example, I'm afraid people will get a false understanding of `depends_on`.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 8]
```diff
@@ -106,5 +113,11 @@ services:
...
       - ACTIVEMQ_BROKER_USERNAME=admin
       - ACTIVEMQ_BROKER_PASSWORD=password
+    depends_on:
+      - mariadb
+      - activemq
     volumes:
       - data:/data
+    
+
+
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Could you remove unneeded newline changes?

**Comment by mtneug** [comment]
If applicable could you adjust `docker-compose.allinone.h2.yml`, `docker-compose.allinone.mariadb.yml`, and `docker-compose.build.yml` as well?

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 2 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Meanwhile, I support that the application itself has to do their own checks, with the sources you give me now I fell betrayed by Docker  😫. The "depends_on" statement loses all their meaning if not can check if a container is ready.

I will delete that lines

**Comment by mliradelc** [comment]
> If applicable could you adjust `docker-compose.allinone.h2.yml`, `docker-compose.allinone.mariadb.yml`, and `docker-compose.build.yml` as well?

No problem, but I will do another PR keep the consistency of the topics.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 9]
```diff
@@ -62,6 +63,9 @@ services:
       - ACTIVEMQ_BROKER_PASSWORD=password
     ports:
       - "8080:8080"
+    depends_on:
+      - mariadb
+      - activemq
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Please remove `depends_on` section.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 10]
```diff
@@ -85,6 +89,9 @@ services:
       - ACTIVEMQ_BROKER_PASSWORD=password
     ports:
       - "8081:8080"
+    depends_on:
+      - mariadb
+      - activemq
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Please remove depends_on section.

File: docker-compose/docker-compose.multiserver.mariadb.yml [hunk 11]
```diff
@@ -107,4 +114,4 @@ services:
       - ACTIVEMQ_BROKER_USERNAME=admin
       - ACTIVEMQ_BROKER_PASSWORD=password
     volumes:
-      - data:/data
+      - data:/data
```
Comment (mtneug) [inline docker-compose/docker-compose.multiserver.mariadb.yml]: Please fix no newline error. [A single newline should end the file](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file/18789#18789).

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 10 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: I see. I amended the code.

File: docker-compose/docker-compose.multiserver.mariadb.yml (hunk 11 above)
Comment (mliradelc) [reply docker-compose/docker-compose.multiserver.mariadb.yml]: Ready

**Comment by mtneug** [comment]
Thank you looks good now. I created #113 for the other files and will merge this forward to `master` and `6.4`.
